monitored_cache.go

```diff
@@ -0,0 +1,151 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cloudsqlconn
+
+import (
+	"context"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"cloud.google.com/go/cloudsqlconn/debug"
+	"cloud.google.com/go/cloudsqlconn/instance"
+	"cloud.google.com/go/cloudsqlconn/internal/cloudsql"
+)
+
+// monitoredCache is a wrapper around a connectionInfoCache that tracks the
+// number of connections to the associated instance.
+type monitoredCache struct {
+	openConnsCount *uint64
+	cn             instance.ConnName
+	resolver       instance.ConnectionNameResolver
+	logger         debug.ContextLogger
+
+	// domainNameTicker periodically checks any domain names to see if they
+	// changed.
+	domainNameTicker *time.Ticker
+	closedCh         chan struct{}
+
+	mu        sync.Mutex
+	openConns []*instrumentedConn
+	closed    bool
+
+	connectionInfoCache
+}
+
+func newMonitoredCache(ctx context.Context, cache connectionInfoCache, cn instance.ConnName, failoverPeriod time.Duration, resolver instance.ConnectionNameResolver, logger debug.ContextLogger) *monitoredCache {
+	c := &monitoredCache{
+		openConnsCount:      new(uint64),
+		closedCh:            make(chan struct{}),
+		cn:                  cn,
+		resolver:            resolver,
+		logger:              logger,
+		connectionInfoCache: cache,
+	}
+	if cn.HasDomainName() {
+		c.domainNameTicker = time.NewTicker(failoverPeriod)
+		go func() {
+			for {
+				select {
+				case <-c.domainNameTicker.C:
+					c.purgeClosedConns()
+					c.checkDomainName(ctx)
+				case <-c.closedCh:
+					return
+				}
+			}
+		}()
+
+	}
+
+	return c
+}
+func (c *monitoredCache) isClosed() bool {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	return c.closed
+}
+
+func (c *monitoredCache) Close() error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if c.closed {
+		return nil
+	}
+
+	c.closed = true
+	close(c.closedCh)
+
+	if c.domainNameTicker != nil {
+		c.domainNameTicker.Stop()
+	}
+
+	if atomic.LoadUint64(c.openConnsCount) > 0 {
+		for _, socket := range c.openConns {
+			if !socket.isClosed() {
+				_ = socket.Close() // force socket closed, ok to ignore error.
+			}
+		}
+		atomic.StoreUint64(c.openConnsCount, 0)
+	}
+
+	return c.connectionInfoCache.Close()
+}
+
+func (c *monitoredCache) ForceRefresh() {
```

Comment:
By embedding the interface above, you get these delegators automatically, so as long as you're just calling through, you don't need these methods.

Reply:
Fixed.